modules/consensus/accept.go

```diff
@@ -207,78 +195,110 @@ func (cs *ConsensusSet) addBlockToTree(b types.Block) (ce changeEntry, err error
-			// a new block to the cache.
-			if err == errFutureTimestamp {
-				go func() {
-					time.Sleep(time.Duration(b.Timestamp-(types.CurrentTimestamp()+types.FutureThreshold)) * time.Second)
-					err := cs.managedAcceptBlock(b)
-					if err != nil {
-						cs.log.Debugln("WARN: failed to accept a future block:", err)
-					}
-					cs.managedBroadcastBlock(b)
-				}()
+		for _, b := range blocks {
+			// Do some relatively inexpensive checks to validate the header and block.
+			// Validation generally occurs in the order of least expensive validation
+			// first.
+			err := cs.validateHeaderAndBlock(boltTxWrapper{tx}, b)
+			if err != nil {
+				// If the block is in the near future, but too far to be acceptable, then
+				// save the block and add it to the consensus set after it is no longer
+				// too far in the future.
+				//
+				// TODO: an attacker could mine many blocks off the genesis block all in the
+				// future and we would spawn a goroutine per each block. To fix this, either
+				// ban peers that send lots of future blocks and stop spawning goroutines
+				// after we are already waiting on a large number of future blocks.
+				//
+				// TODO: an attacker could broadcast a future block many times and we would
+				// spawn a goroutine for each broadcast. To fix this we should create a
+				// cache of future blocks, like we already do for DoS blocks, and only spawn
+				// a goroutine if we haven't already spawned one for that block. To limit
+				// the size of the cache of future blocks, make it a constant size (say 50)
+				// over which we would evict the block furthest in the future before adding
+				// a new block to the cache.
+				if err == errFutureTimestamp {
+					go func() {
+						time.Sleep(time.Duration(b.Timestamp-(types.CurrentTimestamp()+types.FutureThreshold)) * time.Second)
+						err := cs.managedAcceptBlock(b)
+						if err != nil {
+							cs.log.Debugln("WARN: failed to accept a future block:", err)
+						}
+						cs.managedBroadcastBlock(b)
+					}()
+				}
+				if err == modules.ErrNonExtendingBlock || err == modules.ErrBlockKnown {
+					err2 = err
+					continue
+				}
+				return err
 			}
-			return err
+
+			// Try adding the block to the block tree. This call will perform
+			// verification on the block before adding the block to the block tree. An
+			// error is returned if verification fails or if the block does not extend
+			// the longest fork.
+			changeEntry, err := cs.addBlockToTree(tx, b)
+
+			// modules.ErrNonExtendingBlock should be returned if the block does
+			// not extend the current blockchain, however the changes from
+			// addBlockToTree should be committed (which means 'nil' must
+			// be returned). A flag is set to indicate that modules.ErrNonExtending
+			// should be returned.
+			if err == modules.ErrNonExtendingBlock || err == modules.ErrBlockKnown {
+				err2 = err
+				continue
+			} else if err != nil {
+				return err
+			}
```

Comment:
I think that's okay, you'd have to re-download the valid ones from a new peer, but I don't think you'd get stuck.